### PR TITLE
Support Snapshots in CRI

### PR DESCRIPTION
**Pull request by plamenmpetrov**

### CODE REVIEW

File: cri/coordinator.go [hunk 1]
```diff
@@ -63,73 +66,184 @@ func newCoordinator(orch *ctriface.Orchestrator, opts ...coordinatorOption) *coo
...
+		return err
+	}
+
+	fi.logger.Debug("successfully loaded idle instance")
+	return nil
+}
+
+func (c *coordinator) orchCreateSnapshot(ctx context.Context, fi *funcInstance) error {
+	var err error
+	// plamen: set timeout??
```
Comment (ustiugov) [inline cri/coordinator.go:199]: just replicate from functions.go

File: cri/coordinator.go [hunk 2]
```diff
@@ -63,73 +66,184 @@ func newCoordinator(orch *ctriface.Orchestrator, opts ...coordinatorOption) *coo
...
-	return resp, err
+	fi := newFuncInstance(vmID, image, resp)
+	logger.Debug("successfully created fresh instance")
+	return fi, err
 }
 
-func (c *coordinator) orchStopVM(ctx context.Context, vmID string) error {
+func (c *coordinator) orchLoadInstance(ctx context.Context, fi *funcInstance) error {
+	fi.logger.Debug("found idle instance to load")
+	// (plamen) set timeout?
```
Comment (ustiugov) [inline cri/coordinator.go:182]: just replicate from functions.go. put a generous timeout

File: cri/coordinator.go [hunk 3]
```diff
@@ -63,73 +66,184 @@ func newCoordinator(orch *ctriface.Orchestrator, opts ...coordinatorOption) *coo
...
+		},
+	)
+
+	logger.Debug("creating fresh instance")
+
+	var (
+		resp *ctriface.StartVMResponse
+		err  error
+	)
+	// (plamen) set timeout?
```
Comment (ustiugov) [inline cri/coordinator.go:167]: just replicate from functions.go

**Review by ustiugov — CHANGES_REQUESTED** [review]

File: cri/coordinator.go [hunk 4]
```diff
@@ -63,73 +67,195 @@ func newCoordinator(orch *ctriface.Orchestrator, opts ...coordinatorOption) *coo
...
+	fi.logger.Debug("successfully loaded idle instance")
+	return nil
+}
+
+func (c *coordinator) orchCreateSnapshot(ctx context.Context, fi *funcInstance) error {
+	var err error
+
+	fi.onceCreateSnapInstance.Do(
+		func() {
+			ctxTimeout, cancel := context.WithTimeout(ctx, time.Second*10)
```
Comment (ustiugov) [inline cri/coordinator.go:209]: set this to 1 min (e.g., HDD-based snaps)

File: knative_workloads/helloworldSerial.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+apiVersion: serving.knative.dev/v1 # Current version of Knative
```
Comment (ustiugov) [inline knative_workloads/helloworldSerial.yaml:1]: why another helloworld yaml?

**Review by ustiugov — CHANGES_REQUESTED** [review]

File: knative_workloads/helloworldSerial.yaml (hunk 5 above)
Comment (plamenmpetrov) [reply knative_workloads/helloworldSerial.yaml:1]: I have made this one have a concurrency limit of 1. This is so that autoscaling can be tested easily.

**Review by ustiugov — APPROVED** [review]

**Pull request merged**